### PR TITLE
Fix Runtime instance stacking

### PR DESCRIPTION
**Pull request by camilbancioiu**
This PR unifies the state stack and instance stack of the `RuntimeContext`, with the following consequences:
* Code is simplified, because the primitive method `executeSmartContractCall()` will neither push nor pop the Runtime instance stack; this means that the code that calls `executeSmartContractCall()` must handle both state stacking and the remaining gas, after the contract is executed indirectly.
* Fixes a gas-related issue on `ExecuteOnSameContext()` and `ExecuteOnDestContext()`, where `GetPointsUsed()` was called on the wrong Wasmer instance.

### CODE REVIEW

File: arwen/elrondapi/elrondei.go [hunk 1]
```diff
@@ -1732,10 +1647,105 @@ func getOriginalTxHash(context unsafe.Pointer, dataOffset int32) {
...
+			return "", nil, 0, err
+		}
+
+		currOffset += actualLen
+	}
+
+	return string(function), data, currOffset - dataOffset, nil
+}
+
+func bytesToInt32(data []byte) int32 {
```
Comment (andrei-marinica) [inline arwen/elrondapi/elrondei.go:1744]: pretty sure this function can be taken from some library
I think this one does it:
https://golang.org/pkg/encoding/binary/#example_ByteOrder_get

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

File: arwen/elrondapi/elrondei.go [hunk 2]
```diff
@@ -1732,10 +1647,105 @@ func getOriginalTxHash(context unsafe.Pointer, dataOffset int32) {
 	_ = arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution())
 }
 
-func areInSameShard(host arwen.VMHost, leftAddress []byte, rightAddress []byte) bool {
-	blockchain := host.Blockchain()
-	leftShard := blockchain.GetShardOfAddress(leftAddress)
-	rightShard := blockchain.GetShardOfAddress(rightAddress)
+func prepareIndirectContractCallInput(
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1650]: this PR has some changes from the previous one which is not yet merged. let's do them step by step.

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: arwen/elrondapi/elrondei.go (hunk 2 above)
Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go:1650]: Indeed, this depends on https://github.com/ElrondNetwork/arwen-wasm-vm/pull/213 must be merged first.

File: arwen/elrondapi/elrondei.go (hunk 1 above)
Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go:1754]: Changed to use `binary.LittleEndian.Uint32()`.

**Review by andrei-marinica — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**